### DISCUSSION TITLE
Customizable statusbar - multiple separators and colors

### DISCUSSION THREAD

**Discussion by quomat** [helix-editor/helix#10402] [Ideas]
Hi,
currently Helix's statusbar can be configured using a handful of components. There is also a separator support.
The most important component, mode indicator, can be colored using `color-modes = true` in the config.
Themes can set the following colors:
```
ui.statusline.inactive
ui.statusline.normal 
ui.statusline.insert
ui.statusline.select
ui.statusline.separator
```

However, this set of properties is limited. I would expect to be able to customize it like in the nvim plugin [lualine](https://github.com/nvim-lualine/lualine.nvim). Reference images:
<p>
<img width='700' src='https://user-images.githubusercontent.com/41551030/108650373-bb025580-74bf-11eb-8682-2c09321dd18e.png'/>
<img width='700' src='https://user-images.githubusercontent.com/41551030/108650377-bd64af80-74bf-11eb-9c55-fbfc51b39fe8.png'/>
<img width='700' src='https://user-images.githubusercontent.com/41551030/108650378-be95dc80-74bf-11eb-9718-82b242ecdd54.png'/>
<img width='700' src='https://user-images.githubusercontent.com/41551030/108650381-bfc70980-74bf-11eb-9245-85c48f0f154a.png'/>
<img width='700' src='https://user-images.githubusercontent.com/41551030/103467925-32372b00-4d54-11eb-88d6-6d39c46854d8.png'/>
</p>

One particular problem I had is that the powerline separators can not be colored at all, so with the `color-modes = true` the slant lines are not colored. More generally, having the capability to color every component of the statusline would be helpful for theme creators.

Either:
- we wait for the plugin support to make a similar plugin for Helix
- add the statusbar customization to the core Helix

I think that Helix's strong side is that it's amazing without plugins and we already have this small set of statusline coloring properties, so I feel it is not that far of a stretch to extend the customization further.

**Comment by CptPotato** [comment]
For a mode separator with theming support there's already #5022.

**Reply by silverhairs** [reply]
Seems like the PR was discontinued. Is there plans to continue this?

**Comment by usagi-flow** [comment]
I like the idea a lot and don't think this functionality should be a plugin.

This level of customization was a bit out of scope when we made the status line configurable (#2434): There wasn't even a colored mode back then.
However, I definitely believe the suggested configurability belongs to the [existing feature](https://docs.helix-editor.com/configuration.html#editorstatusline-section).

I think it'd be great if we'd generalize the theming of the mode and mode separator, i.e. each element can, optionally, be styled (e.g. like the mode element, but it could also be styled differently).

I imagine a generic and realistic solution as follows:
- We wouldn't just have "colored" and "not colored". For each status line element, we could define a reference to a theme entry.
  - Such references would be included in the existing status line configuration.
  - We could differentiate between a _reference to a static color_, and a _reference to a set of mode colors_.
- We could, by default, keep the existing theme entries, but a user could define extra entries in their theme.
  - This would allow "colored" elements, "normal" elements, but also "emphasized"/"highlighted" (?) elements, just like with Lualine.

Any further feedback to this would be greatly appreciated. :)

**Comment by ItsSunnyMonster** [comment]
It seems that #5022 was closed last year without any reason. Are there still plans to support this?

**Comment by sihaixianyu** [comment]
Are there still plans to support this?

**Comment by Mazrine** [comment]
this would be amazing

**Comment by n6v26r** [comment]
#13666 implements most of this
I would also like to see some formatting options for components (kinda what starship.rs does)

**Reply by Alonely0** [reply]
I don't think it entirely makes sense for helix to have a system like starship.rs' because helix makes a quite hard separation between theming and configuration, and starship.rs has them together with explicit style variables. However, I am more keen on custom formatting with implicit styling, such that the styling is defined in the theme file, while in the configuration we just tweak presentation elements (such as powerline-esque separators). Helix is already an editor that encourages you to define custom themes inheriting from existing ones, so it allows us to position ourselves harder in this philosophy while keeping the statusline configuration surface quite sane; as a +1, this keeps compatibility with existing statusline theming efforts, so we don't have to supersede them.

**Reply by Alonely0** [reply]
I honestly volunteer to implement this if the Helix team decides it's something worth implementing.
